src/libxively/gateway/xi_gw.c

```diff
@@ -0,0 +1,232 @@
+
+        ed_id_context->edge_device_context->context_data.main_context_handle = xih;
+
+        xi_connect_with_lastwill_to_impl(
+            ed_id_context->edge_device_context, "gateway no host", 0,
+            "edge device username", "edge device password", connection_timeout,
+            keepalive_timeout, session_type, will_topic, will_message, will_qos,
+            will_retain, client_callback );
+
+        /*
+         * todo_atigyi:
+         * - error handling
+         * - get corresponding credentials to application device id
+         * - call init on top layer
+         */
+    }
+    else
+    {
+        /* there is already such a context */
+    }
+
+    return state;
+
+err_handling:
+
+    XI_SAFE_FREE( ed_id_context->edge_device_id );
+    XI_SAFE_FREE( ed_id_context );
+
+    return state;
+}
+
+xi_state_t xi_disconnect_ed( xi_context_handle_t xih, const char* edge_device_id )
+{
+    printf( "%s\n", __FUNCTION__ );
+    XI_UNUSED( xih );
+
+    xi_state_t state = XI_STATE_OK;
+
+    xi_ed_id_ed_context_pair_t search_key = {.edge_device_id = ( char* )edge_device_id,
+                                             .edge_device_context = NULL};
+
+    const xi_vector_index_type_t found_index = xi_vector_find(
+        xi_globals.context_handles_vector_edge_devices,
+        XI_VEC_CONST_VALUE_PARAM( ( void* )&search_key ), _xi_ed_find_context_2_id );
+
+    if ( 0 <= found_index )
+    {
+        xi_ed_id_ed_context_pair_t* ed_id_context =
+            ( xi_ed_id_ed_context_pair_t* )xi_vector_get(
+                xi_globals.context_handles_vector_edge_devices, found_index );
+
+        state = xi_shutdown_connection_impl( ed_id_context->edge_device_context );
+    }
+
+    return state;
+}
+
+extern xi_state_t xi_remove_ed( xi_context_handle_t xih, const char* edge_device_id )
+{
+    printf( "%s\n", __FUNCTION__ );
+    XI_UNUSED( xih );
+
+    xi_ed_id_ed_context_pair_t search_key = {.edge_device_id = ( char* )edge_device_id,
+                                             .edge_device_context = NULL};
+
+    const xi_vector_index_type_t found_index = xi_vector_find(
+        xi_globals.context_handles_vector_edge_devices,
+        XI_VEC_CONST_VALUE_PARAM( ( void* )&search_key ), _xi_ed_find_context_2_id );
+
+    if ( 0 <= found_index )
+    {
+        xi_ed_id_ed_context_pair_t* ed_id_context =
+            ( xi_ed_id_ed_context_pair_t* )xi_vector_get(
+                xi_globals.context_handles_vector_edge_devices, found_index );
+
+        xi_delete_context_with_custom_layers(
+            &ed_id_context->edge_device_context, xi_gw_layer_chain,
+            XI_LAYER_CHAIN_SCHEME_LENGTH( XI_LAYER_CHAIN_GW ) );
+
+        xi_vector_del( xi_globals.context_handles_vector_edge_devices, found_index );
+
+        XI_SAFE_FREE( ed_id_context->edge_device_id );
+        XI_SAFE_FREE( ed_id_context );
+    }
+
+    return XI_STATE_OK;
+}
+
+xi_context_handle_t xi_create_gateway_context( xi_context_handle_t context_handle )
+{
```

Comment:
xi_gateway_create_context ?

Reply:
sorry again, this function will be removed completely.

Reply:
👍

Reply:
removed